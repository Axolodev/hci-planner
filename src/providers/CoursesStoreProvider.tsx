import {
  CoursesStoreType,
  createCoursesStore,
  initCounterStore,
} from "@/stores/coursesStore";
import { createContext, ReactNode, useContext, useRef } from "react";
import { useStore } from "zustand";

// Define provides for React
export type CoursesStoreApi = ReturnType<typeof createCoursesStore>;

export const CoursesStoreContext = createContext<CoursesStoreApi | undefined>(
  undefined
);

export interface CoursesStoreProviderProps {
  children: ReactNode;
}

export const CoursesStoreProvider = ({
  children,
}: CoursesStoreProviderProps) => {
  const storeRef = useRef<CoursesStoreApi | null>(null);
  if (storeRef.current === null) {
    storeRef.current = createCoursesStore(initCounterStore());
  }

  return (
    <CoursesStoreContext.Provider value={storeRef.current}>
      {children}
    </CoursesStoreContext.Provider>
  );
};

// Define a hook to access the store
export const useCoursesStore = <T,>(
  selector: (store: CoursesStoreType) => T
): T => {
  const coursesStoreContext = useContext(CoursesStoreContext);

  if (!coursesStoreContext) {
    throw new Error(`useCoursesStore must be used within CoursesStoreProvider`);
  }

  return useStore(coursesStoreContext, selector);
};
