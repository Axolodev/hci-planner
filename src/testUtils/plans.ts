import { Plan } from "@/types";

export const basicPlan = {
  sections: [
    {
      title: "Section title",
      options: [
        {
          name: "Option name",
          modules: [
            {
              name: "Module name",
            },
          ],
        },
      ],
      description: "",
    },
  ],
};

export const planWithMultipleSectionOptions: Plan = {
  sections: [
    {
      title: "Section title",
      options: [
        {
          name: "Option name",
          modules: [
            {
              name: "Module name",
            },
          ],
        },
        {
          name: "Another option",
          modules: [
            {
              name: "Another module",
            },
          ],
        },
      ],
      description: "",
    },
  ],
};
