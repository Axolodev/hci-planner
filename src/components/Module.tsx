import { useCoursesStore } from "@/providers/CoursesStoreProvider";
import { Module as ModuleType } from "@/types";
import * as React from "react";

export const TESTING_ID = "Module";

interface Props extends ModuleType {
  moduleIndex: number;
  optionName: string;
  sectionTitle: string;
  optionIndex: number;
  disabled: boolean;
}

const Module: React.FunctionComponent<Props> = ({
  name,
  sectionTitle,
  optionIndex,
  moduleIndex,
  optionName,
  disabled,
}) => {
  const { getModuleStatus, setModuleStatus } = useCoursesStore(
    (state) => state
  );
  const isCompleted = getModuleStatus(sectionTitle, optionIndex, moduleIndex);
  const handleChange: React.ChangeEventHandler<HTMLInputElement> = (event) => {
    setModuleStatus(
      sectionTitle,
      optionIndex,
      moduleIndex,
      event.target.checked
    );
  };

  return (
    <input
      type="checkbox"
      onChange={handleChange}
      checked={isCompleted}
      disabled={disabled}
      id={`${optionName}-${name}`}
      className={`btn join-item disabled:cursor-not-allowed w-full rounded-field min-h-[40px] h-auto`}
      aria-label={name}
      style={{ margin: 0 }}
      data-testid={TESTING_ID}
    />
  );
};

export default Module;
