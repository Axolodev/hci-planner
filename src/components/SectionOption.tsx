import { SectionOption as SectionOptionType } from "@/types";
import Module from "./Module";
import { useCoursesStore } from "@/providers/CoursesStoreProvider";

interface Props {
  option: SectionOptionType;
  sectionTitle: string;
  optionIndex: number;
}

export default function SectionOption({
  option,
  sectionTitle,
  optionIndex,
}: Props) {
  const {
    getIsNeighbourSectionOptionStarted: isNeighbourSectionOptionStarted,
  } = useCoursesStore((state) => state);
  const isNeighborStarted = isNeighbourSectionOptionStarted(
    sectionTitle,
    optionIndex
  );

  return (
    <div
      className={`p-4 flex-1 rounded-box flex flex-col items-stretch gap-2 justify-center transition-colors join w-full ${
        isNeighborStarted ? "bg-base-300/60 cursor-not-allowed" : "bg-base-300"
      }`}
    >
      {option.modules.map((module, index) => (
        <Module
          key={module.name}
          name={module.name}
          optionName={option.name}
          moduleIndex={index}
          sectionTitle={sectionTitle}
          optionIndex={optionIndex}
          disabled={isNeighborStarted}
        />
      ))}
    </div>
  );
}
