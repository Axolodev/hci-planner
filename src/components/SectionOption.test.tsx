import React from "react";
import { render } from "@testing-library/react";
import { CoursesStoreProvider } from "../providers/CoursesStoreProvider";
import { basicPlan, planWithMultipleSectionOptions } from "../testUtils/plans";
import { TESTING_ID as MODULE_TESTING_ID } from "./Module";
import SectionOption, {
  TESTING_ID_PREFIX as SECTION_OPTION_TESTING_ID_PREFIX,
} from "./SectionOption";
import userEvent from "@testing-library/user-event";

describe("SectionOption", () => {
  it("renders correctly and matches snapshot", () => {
    const { container } = render(
      <CoursesStoreProvider plan={basicPlan}>
        <SectionOption
          option={basicPlan.sections[0].options[0]}
          sectionTitle="Section title"
          optionIndex={0}
        />
      </CoursesStoreProvider>
    );
    expect(container).toMatchSnapshot();
  });

  it("can check modules within a sectionOption", async () => {
    const { getByTestId } = render(
      <CoursesStoreProvider plan={basicPlan}>
        <SectionOption
          option={basicPlan.sections[0].options[0]}
          sectionTitle="Section title"
          optionIndex={0}
        />
      </CoursesStoreProvider>
    );
    const checkbox = getByTestId(MODULE_TESTING_ID) as HTMLInputElement;
    expect(checkbox.checked).toBe(false);
    await userEvent.click(checkbox);
    expect(checkbox.checked).toBe(true);
  });

  it("should disable sectionOption when another is selected", async () => {
    const { getByTestId } = render(
      <CoursesStoreProvider plan={planWithMultipleSectionOptions}>
        <SectionOption
          option={planWithMultipleSectionOptions.sections[0].options[0]}
          sectionTitle="Section title"
          optionIndex={0}
        />
        <SectionOption
          option={planWithMultipleSectionOptions.sections[0].options[1]}
          sectionTitle="Section title"
          optionIndex={1}
        />
      </CoursesStoreProvider>
    );
    const firstCheckbox = getByTestId(
      `${SECTION_OPTION_TESTING_ID_PREFIX}0`
    ).querySelector("input") as HTMLInputElement;
    const secondCheckbox = getByTestId(
      `${SECTION_OPTION_TESTING_ID_PREFIX}1`
    ).querySelector("input") as HTMLInputElement;
    expect(firstCheckbox.disabled).toBe(false);
    expect(secondCheckbox.disabled).toBe(false);
    await userEvent.click(firstCheckbox);
    expect(firstCheckbox.disabled).toBe(false);
    expect(secondCheckbox.disabled).toBe(true);
  });
});
