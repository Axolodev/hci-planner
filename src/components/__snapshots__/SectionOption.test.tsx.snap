// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SectionOption renders correctly and matches snapshot 1`] = `
<div>
  <div
    class="p-4 flex-1 rounded-box flex flex-col items-stretch gap-2 justify-center transition-colors join w-full bg-base-300"
  >
    <input
      aria-label="Module name"
      class="btn join-item disabled:cursor-not-allowed w-full rounded-field"
      data-testid="Module"
      id="Option name-Module name"
      style="margin: 0px;"
      type="checkbox"
    />
  </div>
</div>
`;
